kernel/rtlil.h

```diff
@@ -1501,6 +1509,31 @@ struct RTLIL::Module : public RTLIL::AttrObject
 #endif
 };
 
+struct RTLIL::Function : public RTLIL::AttrObject
+{
+	unsigned int hashidx_;
+	unsigned int hash() const { return hashidx_; }
+
+protected:
+	// use module->addWire() and module->remove() to create or destroy wires
+	friend struct RTLIL::Module;
+	Function();
+	~Function();
+
+public:
+	// do not simply copy wires
+	Function(RTLIL::Function &other) = delete;
+	void operator=(RTLIL::Function &other) = delete;
+
+	RTLIL::Module *module;
+	RTLIL::IdString name;
+	AST::AstNode *def;
```

Comment:
As far as I can tell this new `Function` object is here so that the ASTs of functions can be looked up. What is the AST used for?

This is adding a new type of member to modules (next to wires, cells, processes and memories). This means these now have to be considered all over the Yosys codebase anywhere we are manipulating modules and they should be saved/restored on writing out an RTLIL file.

That's a substantial addition in complexity to serve one feature. Preferably we encode what we need in parameters of the new cell type `$fun` instead (if we need a new cell type in the first place).